### ISSUE TITLE
ohmyzsh is preventing updated midnight commander to switch to shell

### ISSUE THREAD

**Issue by andyrozman** [ohmyzsh/ohmyzsh#11353]
### Describe the bug

In ubuntu 22.04 ohmyzsh is preventing midnight commander to switch to shell (Turn off panels as command is called in mc). So whenever I try to switch to shell (Ctrl+o) I just get returned back. This worked correctly in all previous LTS versions (20.04, 18.04) of ubuntu (I always use LTS, no other versions in between), but stopped working in this version.

I have found a workaround, but makes my expierence with ohmyzsh and mc less than good (dasabling use of zsh and ohmyzsh within mc).
Workaround: define alias that starts mc with bash:
alias mc="SHELL=/bin/bash TERM=xterm-256color /usr/bin/mc"

### Steps to reproduce

1. Install Ubuntu 22.04
2. Install ohmyzsh
3. Run midnight commander (mc)
4. Switch to console / Turn panels off (Ctrl+o)

### Expected behavior

Shell console is shown.

Current behavior: we get returned to mc at once

### OS / Linux distribution

Ubuntu 22.04 LTS

### Zsh version

5.8.1

### Terminal emulator

Terminator

**Comment by mcornella** [comment]
Can't reproduce this on ubuntu:jammy docker image with mc 4.8.27. Could you capture a trace of your startup files and post them here or send them at logs@ohmyz.sh?

You can take a capture of that with `zsh -xvic exit &>~/omz-debug.log` and posting the `omz-debug.log` file.

**Comment by mcornella** [comment]
I did not see any relevant culprit in that trace. Try instead the following:
1. Open two terminals
2. In one of the terminals, add the code below at the top of the .zshrc file:
   ```sh
   exec > >(tee ~/omz-debug.log) 2>&1
   set -xv
   ```
4. In the other terminal, open mc and switch to console, and wait for the terminal to close back.
5. In the first terminal, remove the lines added to the .zshrc file.
6. Send the log file to logs@ohmyz.sh as a reply to the email you sent.
Thank you for your collaboration!

**Comment by lazystone** [comment]
@mcornella in case if you haven't got logs, here is a relevant part of it:
```
mc
mc
+omz_termsupport_preexec:1> [[ '' != true ]]
+omz_termsupport_preexec:3> emulate -L zsh
+omz_termsupport_preexec:4> setopt extended_glob
+omz_termsupport_preexec:7> local -a cmdargs
+omz_termsupport_preexec:8> cmdargs=( mc ) 
+omz_termsupport_preexec:10> [[ mc = fg ]]
+omz_termsupport_preexec:44> local CMD=mc
+omz_termsupport_preexec:45> local LINE=mc
+omz_termsupport_preexec:47> title mc '%100>...>mc%<<'
+title:1> setopt localoptions nopromptsubst
+title:4> [[ -n '' ]]
+title:8> : '%100>...>mc%<<'
+title:10> case xterm-256color (cygwin | xterm*)
+title:12> print -Pn '\e]2;%100\>...\>mc%\<\<\a'
+title:13> print -Pn '\e]1;mc\a'
+/usr/bin/zsh:1> mc
Cannot get terminal settings: Inappropriate ioctl for device (25)
```

Also emailed full log to the address from previous message.

Funny thing that on another laptop(same ubuntu version) everything works, so I suspect that it could be some plugins/settings..

**Comment by mcornella** [comment]
It looks like stdout is not initialized as a tty, sample from the beginning of the init process:
```shell
# - is not run from a tty
# - git is unavailable on the system.
if [[ "$update_mode" = disabled ]] \
   || [[ ! -w "$ZSH" || ! -O "$ZSH" ]] \
   || [[ ! -t 1 ]] \
   || ! command git --version 2>&1 >/dev/null; then
  unset update_mode
  return
fi
+/home/andrej/.oh-my-zsh/tools/check_for_upgrade.sh:26> [[ prompt = disabled ]]
+/home/andrej/.oh-my-zsh/tools/check_for_upgrade.sh:27> [[ ! -w /home/andrej/.oh-my-zsh || ! -O /home/andrej/.oh-my-zsh ]]
+/home/andrej/.oh-my-zsh/tools/check_for_upgrade.sh:28> [[ ! -t 1 ]]
+/home/andrej/.oh-my-zsh/tools/check_for_upgrade.sh:30> unset update_mode
```

The `[ -t 1 ]` check fails. This is why you get the error
```
Cannot get terminal settings: Inappropriate ioctl for device (25)
```

**Comment by andyrozman** [comment]
Any solution to this? 

My system worked correctly before I upgraded to Ubuntu 22.04 LTS (from 20.04), and after that midnight commander seems to be only application that doesn't work correctly... My workaround works, but it is anoyying not be able to use zsh in shell under mc.

**Comment by carlosala** [comment]
I was not able to reproduce on ubuntu 22.04 with default settings. Could anyone share a .zshrc that reproduces it?

**Comment by lazystone** [comment]
I found one thing playing with this I could reproduce/fix this issue.

I use [podman](https://podman.io/docs/installation) instead of docker and in my .zshrc I have `export DOCKER_HOST="unix://$(podman info -f "{{.Host.RemoteSocket.Path}}")"` and shell in MC does not work then.

If I comment this out or replace by `export DOCKER_HOST=unix:///run/user/1000/podman/podman.sock` then shell in MC works.

**Comment by mcornella** [comment]
Can you test whether commenting out all zshrc content works, and then if adding `sleep 2` makes it not work again? I think https://github.com/MidnightCommander/mc/commit/e89713a3d6a2edd4813958608ce461540c4f715e might be at fault here.

**Comment by lazystone** [comment]
Ok, I got this issue again. MC version is 4.8.27

Then I upgraded to the latest available in ppa: 4.8.30 - issue disappeared.

@mcornella tried to add `sleep 2` to zshrc - nope, still works(4.8.30).

**Comment by xvoronina** [comment]
I was able to reproduce the issue on Ubuntu 22.04, mc 4.8.30. I had the following string in .zshrc
"alias mc="mc --nosubshell"" to stop  mc from lugging when it starts in zsh. I removed that string and reload zsh. It did not help. I re-installed mc 4.8.30 and it seems to work now if I want to switch to shelll (ctrl+o) but sorrowly mc luggs on start up.

**Comment by mcornella** [comment]
@xvoronina so just to confirm, this is solved with 4.8.30 or the `--nosubshell` option is needed?

**Comment by NorthLight-EWR** [comment]
I read all your posts, but got somehow lost.
- I have a default ubuntu with latest lts (22.04.3 LTS).
- mc is v4.8.27
- OMZ is latest
- PowerLevel10K

I am not exactly sure, if it is the same, so here's my kind of shell problem in MC:

1. Open mc
2. Strg+O
3. The mc window dissapears to the shell, but i don't get a prompt  as before
4. if i enter _any_ key, i am back in the mc main window
5. Shell not accessible so far

I have tried "--nosubshell", has no impact to that bug.

**Comment by mcornella** [comment]
Here's a summary of what we found:

It appears this issue is related https://github.com/MidnightCommander/mc/commit/e89713a3d6a2edd4813958608ce461540c4f715e

To confirm, you can reproduce it with the following steps:
- Back up you `.zshrc` file somewhere, and make it a blank file with no contents. Issue should be solved
- Then, add `sleep 2` in the `.zshrc`. This should slow down the shell startup enough to make mc skip it.

This is not reproducible on 4.8.30, which should contain the fixes.

TL;DR: update `mc` to a 4.8.30 or newer. I'm going to close this unless someone has a reproducible setup that makes 4.8.30 fail.

**Comment by flocsy** [comment]
I have mc 4.8.31 and still have the probem. In fact I had this problem fixed in the past, and then it came back a few days ago when I ran `brew upgrade`

**Comment by emtiu** [comment]
I just experienced the same problem with mc 4.8.32 on FreeBSD in bash. It's really annoying, but since I never use the shell, I can just disable it by usind `mc -u`.

Leaving this comment to stress that the issue is still present.
